lite/dbprovider.go

```diff
@@ -0,0 +1,260 @@
+	if err != nil {
+		return err
+	}
+	batch.Set(shKey, shBz)
+
+	// And write sync.
+	batch.WriteSync()
+
+	// Garbage collect.
+	// TODO: optimize later.
+	if dbp.limit > 0 {
+		dbp.deleteAfterN(fc.ChainID(), dbp.limit)
+	}
+
+	return nil
+}
+
+// Implements Provider.
+func (dbp *DBProvider) LatestFullCommit(chainID string, minHeight, maxHeight int64) (
+	FullCommit, error) {
+
+	dbp.logger.Info("DBProvider.LatestFullCommit()...",
+		"chainID", chainID, "minHeight", minHeight, "maxHeight", maxHeight)
+
+	if minHeight <= 0 {
+		minHeight = 1
+	}
+	if maxHeight == 0 {
+		maxHeight = 1<<63 - 1
+	}
+
+	itr := dbp.db.ReverseIterator(
+		signedHeaderKey(chainID, maxHeight),
+		signedHeaderKey(chainID, minHeight-1),
+	)
+	defer itr.Close()
+
+	for itr.Valid() {
+		key := itr.Key()
+		_, _, ok := parseSignedHeaderKey(key)
+		if !ok {
+			// Skip over other keys.
+			itr.Next()
+			continue
+		} else {
+			// Found the latest full commit signed header.
+			shBz := itr.Value()
+			sh := types.SignedHeader{}
+			err := dbp.cdc.UnmarshalBinary(shBz, &sh)
+			if err != nil {
+				return FullCommit{}, err
+			} else {
+				lfc, err := dbp.fillFullCommit(sh)
+				if err == nil {
+					dbp.logger.Info("DBProvider.LatestFullCommit() found latest.", "height", lfc.Height())
+					return lfc, nil
+				} else {
+					dbp.logger.Info("DBProvider.LatestFullCommit() got error", "lfc", lfc)
+					dbp.logger.Info(fmt.Sprintf("%+v", err))
+					return lfc, err
+				}
+			}
+		}
+	}
+	return FullCommit{}, lerr.ErrCommitNotFound()
+}
+
+func (dbp *DBProvider) ValidatorSet(chainID string, height int64) (valset *types.ValidatorSet, err error) {
+	return dbp.getValidatorSet(chainID, height)
+}
+
+func (dbp *DBProvider) getValidatorSet(chainID string, height int64) (valset *types.ValidatorSet, err error) {
+	vsBz := dbp.db.Get(validatorSetKey(chainID, height))
+	if vsBz == nil {
+		err = lerr.ErrMissingValidators(chainID, height)
+		return
+	}
+	err = dbp.cdc.UnmarshalBinary(vsBz, &valset)
+	if err != nil {
+		return
+	}
+	valset.TotalVotingPower() // to test deep equality.
+	return
+}
+
+func (dbp *DBProvider) fillFullCommit(sh types.SignedHeader) (FullCommit, error) {
+	var chainID = sh.ChainID
+	var height = sh.Height
+	var valset, nextValset *types.ValidatorSet
+	// Load the validator set.
+	valset, err := dbp.getValidatorSet(chainID, height)
+	if err != nil {
+		return FullCommit{}, err
+	}
+	// Load the next validator set.
+	nextValset, err = dbp.getValidatorSet(chainID, height+1)
+	if err != nil {
+		return FullCommit{}, err
+	}
+	// Return filled FullCommit.
+	return FullCommit{
+		SignedHeader:   sh,
+		Validators:     valset,
+		NextValidators: nextValset,
+	}, nil
+}
+
+func (dbp *DBProvider) deleteAfterN(chainID string, after int) error {
+
+	dbp.logger.Info("DBProvider.deleteAfterN()...", "chainID", chainID, "after", after)
+
+	itr := dbp.db.ReverseIterator(
+		signedHeaderKey(chainID, 1<<63-1),
+		signedHeaderKey(chainID, 0),
```

Comment:
Should it be after instead of 0?

Reply:
no, we want to iterate starting from the latest and go as far back as possible until we find "after" items.  e.g. we might need to return height 1